### PR TITLE
feat: commonjs support import interop function at rust side

### PR DESCRIPTION
**Pull request by underfin**
**Description:**

**BREAKING CHANGE:**

**Related issue (if exists):**
https://github.com/swc-project/swc/issues/7130

### CODE REVIEW

File: crates/swc_ecma_preset_env/data/core-js-compat/data.json
Comment (kdy1) [inline crates/swc_ecma_preset_env/data/core-js-compat/data.json:1]: Please revert this

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/swc_ecma_transforms_module/src/common_js.rs [hunk 1]
```diff
@@ -21,11 +21,14 @@ use crate::{
...
 };
 
+pub type ImportInteropFn = Option<Box<dyn for<'aa> Fn(Option<&'aa str>) -> ImportInterop>>;
+
 pub fn common_js<C>(
     unresolved_mark: Mark,
     config: Config,
     available_features: FeatureFlag,
     comments: Option<C>,
+    import_interop_fn: ImportInteropFn,
```
[the remark this answers is not in the document]
Comment (underfin) [reply crates/swc_ecma_transforms_module/src/common_js.rs:31]: Add it to `Config` will need impl trait for `Debug/Clone`, look like it is not good.

[the remark this answers is not in the document]
Comment (kdy1) [reply crates/swc_ecma_transforms_module/src/common_js.rs:31]: Oh, I see.

**Comment by kdy1** [comment]
No worries, it's automatic 😄 .
All changes in the main swc repository are published each time a PR is merged

**Comment by kdy1** [comment]
I discussed about this with the team, and we concluded not to accept this change. I'm sorry, but I want to be more careful about adding apis. (And I hate breaking changes)

**Comment by underfin** [comment]
`Please move this field to config with #[serde(skip_deserializing)]`

If add `import_interop_fn` to `Config` without api change, maybe you like it?

**Comment by kdy1** [comment]
No, it's still an API change.

**Comment by underfin** [comment]
What do you think about this?
```
pub enum ImportInterop {
    #[serde(alias = "babel")]
    Swc,
    Node,
    None,
    Fn(ImportInterOpFn)
}
```

**Comment by kwonoj** [comment]
Debate in here is if we want / expose fn, or another object as an additional way to current api surface, How is the second problem. So far we only had 1 feature request, where we are not able to determine if it is being widely used among swc users, or either there are strong needs that resolves wide ecosystem problem. The original issue does not addresses neither.

Since last year, along with the sdk (swc_core) and its consumer (swc/core, next-swc) we are trying to move towards stable, manageable narrow api surface to control over. There are some historical decisions made early which made swc's current state bit complex - that's something we want to reduce but also not easy, however adding  / expanding new feature / api surface is something we can control & we are trying to do now.

**Pull request closed without merging**